{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-05T04:23:27.970402Z",
     "start_time": "2018-03-05T04:23:27.965367Z"
    }
   },
   "outputs": [],
   "source": [
    "import importlib\n",
    "import os\n",
    "import sys\n",
    "from warnings import warn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-05T04:23:27.981628Z",
     "start_time": "2018-03-05T04:23:27.974254Z"
    }
   },
   "outputs": [],
   "source": [
    "github_kwatme_directory_path = '..'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2018-03-05T04:23:32.604490Z",
     "start_time": "2018-03-05T04:23:27.984223Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 classify\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/classify/notebook/classify.ipynb\n",
      "\tfrom classify import train_and_classify\n",
      "\tfrom classify import __init__\n",
      "2 cluster\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/cluster/notebook/cluster.ipynb\n",
      "\tfrom cluster import nmf_consensus_cluster_with_multiple_k\n",
      "\tfrom cluster import nmf_consensus_cluster\n",
      "\tfrom cluster import __init__\n",
      "\tfrom cluster import hierarchical_consensus_cluster_with_multiple_k\n",
      "\tfrom cluster import count_coclustering_and_normalize\n",
      "\tfrom cluster import hierarchical_consensus_cluster\n",
      "3 context\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/context/notebook/context.ipynb\n",
      "\tfrom context import make_context_matrix\n",
      "\tfrom context import fit_skew_t_pdf\n",
      "\tfrom context import make_context_matrix_and_summarize_context_by_row\n",
      "\tfrom context import __init__\n",
      "\tfrom context import plot_context\n",
      "\tfrom context import fit_skew_t_pdfs\n",
      "\tfrom context import compute_context\n",
      "4 cross_validate\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/cross_validate/notebook/cross_validate.ipynb\n",
      "\tfrom cross_validate import __init__\n",
      "\tfrom cross_validate import cross_validate\n",
      "5 cryptograph\n",
      "\t../cryptograph/notebook doesn't exist.\n",
      "\tfrom cryptograph import access_passcode\n",
      "\tfrom cryptograph import __init__\n",
      "\tfrom cryptograph import crypt_with_aes\n",
      "\tfrom cryptograph import hash_\n",
      "6 dimension_scale\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/dimension_scale/notebook/dimension_scale.ipynb\n",
      "\tfrom dimension_scale import mds\n",
      "\tfrom dimension_scale import __init__\n",
      "7 feature\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/feature/notebook/feature.ipynb\n",
      "\tfrom feature import get_gff3_attribute\n",
      "\tfrom feature import index_gff3_df_by_name\n",
      "\tfrom feature import __init__\n",
      "\tfrom feature import read_gff3_gz\n",
      "\tfrom feature import FeatureHDF5\n",
      "8 gct_gmt\n",
      "\t../gct_gmt/notebook doesn't exist.\n",
      "\tfrom gct_gmt import access_gmt\n",
      "\tfrom gct_gmt import __init__\n",
      "\tfrom gct_gmt import access_gct\n",
      "9 genome\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/genome/notebook/genome.ipynb\n",
      "\tfrom genome import __init__\n",
      "\tfrom genome import Genome\n",
      "10 geo\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/geo/notebook/geo.ipynb\n",
      "\tfrom geo import __init__\n",
      "\tfrom geo import get_and_parse_geo_data\n",
      "11 gsea\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/gsea/notebook/gsea.ipynb\n",
      "\tfrom gsea import compute_gene_scores\n",
      "\tfrom gsea import single_sample_gseas\n",
      "\tfrom gsea import __init__\n",
      "\tfrom gsea import plot_mountain_plot\n",
      "\tfrom gsea import single_sample_gsea\n",
      "\tfrom gsea import gsea\n",
      "12 hdf5\n",
      "\t../hdf5/notebook doesn't exist.\n",
      "\tfrom hdf5 import __init__\n",
      "\tfrom hdf5 import read_where_and_map_column_names\n",
      "13 information\n",
      "\t../information/notebook doesn't exist.\n",
      "\tfrom information import __init__\n",
      "\tfrom information import compute_information_coefficient\n",
      "\tfrom information import compute_entropy\n",
      "\tfrom information import compute_information_distance\n",
      "14 kernel_density\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/kernel_density/notebook/kernel_density.ipynb\n",
      "\tfrom kernel_density import kde\n",
      "\tfrom kernel_density import __init__\n",
      "\tfrom kernel_density import compute_bandwidths\n",
      "15 linear_algebra\n",
      "\t../linear_algebra/notebook doesn't exist.\n",
      "\tfrom linear_algebra import __init__\n",
      "\tfrom linear_algebra import solve_ax_equal_b\n",
      "16 linear_model\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/linear_model/notebook/linear_model.ipynb\n",
      "\tfrom linear_model import correlate\n",
      "\tfrom linear_model import __init__\n",
      "17 match\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/match/notebook/match.ipynb\n",
      "\tfrom match import plot_match_panel\n",
      "\tfrom match import make_summary_match_panel\n",
      "\tfrom match import __init__\n",
      "\tfrom match import make_match_panel\n",
      "\tfrom match import make_comparison_panel\n",
      "\tfrom match import match\n",
      "18 matrix_factorization\n",
      "../matrix_factorization/notebook/matrix_factorization.ipynb doesn't exist.\n",
      "\tfrom matrix_factorization import solve_for_nmf_h\n",
      "\tfrom matrix_factorization import __init__\n",
      "\tfrom matrix_factorization import nmf\n",
      "19 mutational_signature\n",
      "\t../mutational_signature/notebook doesn't exist.\n",
      "\tfrom mutational_signature import __init__\n",
      "\tfrom mutational_signature import mutational_signature\n",
      "\tfrom mutational_signature import loadmutsigs\n",
      "20 nd_array\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/nd_array/notebook/nd_array.ipynb\n",
      "\tfrom nd_array import get_1d_array_unique_objects_in_order\n",
      "\tfrom nd_array import check_nd_array_for_bad_value\n",
      "\tfrom nd_array import compute_log_ratio_between_2_1d_arrays\n",
      "\tfrom nd_array import normalize_2d_array\n",
      "\tfrom nd_array import compute_nd_array_margin_of_error\n",
      "\tfrom nd_array import drop_bad_value_and_apply_function_on_2_1d_arrays\n",
      "\tfrom nd_array import compute_empirical_p_value\n",
      "\tfrom nd_array import cluster_1d_arrays\n",
      "\tfrom nd_array import make_index_and_fraction_grid_coordinates_pair\n",
      "\tfrom nd_array import __init__\n",
      "\tfrom nd_array import apply_function_on_2_2d_arrays_slices\n",
      "\tfrom nd_array import get_intersections_between_2_1d_arrays\n",
      "\tfrom nd_array import clip_nd_array_by_standard_deviation\n",
      "\tfrom nd_array import shuffle_each_2d_array_slice\n",
      "\tfrom nd_array import make_nd_grid_coordinates\n",
      "\tfrom nd_array import cluster_2d_array_slices_by_group\n",
      "\tfrom nd_array import apply_function_on_2_1d_arrays_and_compute_empirical_p_value\n",
      "\tfrom nd_array import normalize_1d_array\n",
      "\tfrom nd_array import compute_1d_array_cumulative_sum\n",
      "\tfrom nd_array import compute_empirical_p_values_and_fdrs\n",
      "\tfrom nd_array import get_coordinates_for_reflection\n",
      "\tfrom nd_array import shift_and_log_nd_array\n",
      "21 network\n",
      "\t../network/notebook doesn't exist.\n",
      "\tfrom network import __init__\n",
      "22 nmf_map\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/nmf_map/notebook/nmf_map.ipynb\n",
      "\tfrom nmf_map import __init__\n",
      "\tfrom nmf_map import anneal_node_and_element_positions\n",
      "\tfrom nmf_map import get_triangulation_edges\n",
      "\tfrom nmf_map import NMFMap\n",
      "\tfrom nmf_map import plot_nmf_map\n",
      "23 plot\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/plot/notebook/plot.ipynb\n",
      "\tfrom plot import plot_box_or_violin\n",
      "\tfrom plot import style\n",
      "\tfrom plot import plot_points\n",
      "\tfrom plot import make_colorscale\n",
      "\tfrom plot import __init__\n",
      "\tfrom plot import plot_distributions\n",
      "\tfrom plot import make_random_hex_color\n",
      "\tfrom plot import plot_and_save\n",
      "\tfrom plot import plot_bar\n",
      "\tfrom plot import plot_table\n",
      "\tfrom plot import plot_heatmap\n",
      "24 regress\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/regress/notebook/regress.ipynb\n",
      "\tfrom regress import train_and_regress\n",
      "\tfrom regress import __init__\n",
      "25 sequence\n",
      "\t../sequence/notebook doesn't exist.\n",
      "\tfrom sequence import process_sequence\n",
      "\tfrom sequence import get_chromosome_size_from_fasta_gz\n",
      "\tfrom sequence import __init__\n",
      "\tfrom sequence import get_sequence_from_fasta_gz\n",
      "26 sequencing_process\n",
      "../sequencing_process/notebook/sequencing_process.ipynb doesn't exist.\n",
      "\tfrom sequencing_process import process_fastq_gz\n",
      "\tfrom sequencing_process import simulate\n",
      "\tfrom sequencing_process import __init__\n",
      "\tfrom sequencing_process import bgzip_and_tabix\n",
      "\tfrom sequencing_process import download_clinvar_vcf_gz\n",
      "\tfrom sequencing_process import make_reference_genome\n",
      "\tfrom sequencing_process import process_vcf_gz\n",
      "\tfrom sequencing_process import process_bam\n",
      "27 spro\n",
      "\t../spro/notebook doesn't exist.\n",
      "\tfrom spro import run\n",
      "\tfrom spro import create\n",
      "\tfrom spro import download\n",
      "\tfrom spro import __init__\n",
      "\tfrom spro import list_\n",
      "\tfrom spro import share\n",
      "\tfrom spro import uninstall\n",
      "\tfrom spro import install\n",
      "\tfrom spro import enter\n",
      "\tfrom spro import get\n",
      "28 support\n",
      "\t../support/notebook doesn't exist.\n",
      "\tfrom support import df\n",
      "\tfrom support import git\n",
      "\tfrom support import iterable\n",
      "\tfrom support import log\n",
      "\tfrom support import conda\n",
      "\tfrom support import exit_\n",
      "\tfrom support import dict_\n",
      "\tfrom support import compression\n",
      "\tfrom support import series\n",
      "\tfrom support import __init__\n",
      "\tfrom support import volume\n",
      "\tfrom support import machine\n",
      "\tfrom support import subprocess_\n",
      "\tfrom support import network\n",
      "\tfrom support import python\n",
      "\tfrom support import multiprocess\n",
      "\tfrom support import json_\n",
      "\tfrom support import str_\n",
      "\tfrom support import path\n",
      "29 tcga\n",
      "\t../tcga/notebook doesn't exist.\n",
      "\tfrom tcga import __init__\n",
      "\tfrom tcga import access_analysis_result\n",
      "30 variant\n",
      "\thttp://localhost:1111/notebooks/Jumis/github_kwatme/variant/notebook/variant.ipynb\n",
      "\tfrom variant import __init__\n",
      "\tfrom variant import access_vcf\n",
      "\tfrom variant import access_vcf_dict\n",
      "\tfrom variant import VariantHDF5\n",
      "\tfrom variant import access_maf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/k/Jumis/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: UserWarning:\n",
      "\n",
      "../gct_gmt/README.md: GCT GMT != Gct Gmt\n",
      "\n",
      "/Users/k/Jumis/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: UserWarning:\n",
      "\n",
      "../geo/README.md: GEO != Geo\n",
      "\n",
      "/Users/k/Jumis/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: UserWarning:\n",
      "\n",
      "../gsea/README.md: GSEA != Gsea\n",
      "\n",
      "/Users/k/Jumis/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: UserWarning:\n",
      "\n",
      "../hdf5/README.md: HDF5 != Hdf5\n",
      "\n",
      "/Users/k/Jumis/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: UserWarning:\n",
      "\n",
      "../nd_array/README.md: ND Array != Nd Array\n",
      "\n",
      "/Users/k/Jumis/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: UserWarning:\n",
      "\n",
      "../nmf_map/README.md: NMF Map != Nmf Map\n",
      "\n",
      "/Users/k/Jumis/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: UserWarning:\n",
      "\n",
      "../spro/README.md: SPro != Spro\n",
      "\n",
      "/Users/k/Jumis/conda/lib/python3.6/site-packages/ipykernel_launcher.py:20: UserWarning:\n",
      "\n",
      "../tcga/README.md: TCGA != Tcga\n",
      "\n"
     ]
    }
   ],
   "source": [
    "n = 0\n",
    "\n",
    "for name in sorted(os.listdir(github_kwatme_directory_path)):\n",
    "\n",
    "    if name in ('.DS_Store', '.idea', '.ipynb_checkpoints', 'environment',\n",
    "                'jupyter_gui', 'kwatme.com', 'probability'):\n",
    "        continue\n",
    "\n",
    "    n += 1\n",
    "    print('{} {}'.format(n, name))\n",
    "\n",
    "    sys.path.insert(0, os.path.join(github_kwatme_directory_path, name))\n",
    "\n",
    "    readme_file_path = os.path.join(github_kwatme_directory_path, name,\n",
    "                                    'README.md')\n",
    "    with open(readme_file_path) as file:\n",
    "\n",
    "        header_title = file.readline().strip()[2:]\n",
    "        title = name.title().replace('_', ' ')\n",
    "        if header_title != title:\n",
    "            warn('{}: {} != {}'.format(readme_file_path, header_title, title))\n",
    "            if header_title.lower() != title.lower():\n",
    "                raise ValueError('{}: {} != {}'.format(readme_file_path,\n",
    "                                                       header_title.lower(),\n",
    "                                                       title.lower()))\n",
    "\n",
    "    notebook_directory_path = os.path.join(github_kwatme_directory_path, name,\n",
    "                                           'notebook')\n",
    "\n",
    "    if os.path.isdir(notebook_directory_path):\n",
    "\n",
    "        notebook_file_path = os.path.join(notebook_directory_path,\n",
    "                                          name + '.ipynb')\n",
    "        if os.path.isfile(notebook_file_path):\n",
    "            print('\\t{}'.format(\n",
    "                notebook_file_path.replace(\n",
    "                    '..',\n",
    "                    'http://localhost:1111/notebooks/Jumis/github_kwatme')))\n",
    "        else:\n",
    "            print('{} doesn\\'t exist.'.format(notebook_file_path))\n",
    "\n",
    "    else:\n",
    "        print('\\t{} doesn\\'t exist.'.format(notebook_directory_path))\n",
    "\n",
    "    for file_name in os.listdir(\n",
    "            os.path.join(github_kwatme_directory_path, name, name)):\n",
    "\n",
    "        if file_name.endswith('.py'):\n",
    "\n",
    "            command = 'from {} import {}'.format(name, file_name[:-3])\n",
    "            print('\\t' + command)\n",
    "            exec(command)\n",
    "\n",
    "    sys.path.pop(0)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
